File: locks/.vendor/.zed/oresoftware/ores-locks-and-leases/src/dart/lib/src/plan.dart
/// Which coordination layers a routine engages. Both false is a deliberate
/// pass-through for tests and single-writer development.
final class LockLayers {
  /// The fiducia-cloud lease: outermost, cross-host, TTL-bounded, fenced.
  final bool fiducia;

  /// A Postgres advisory lock: inner, single-database, released by Postgres.
  final bool pgAdvisory;

  const LockLayers({required this.fiducia, required this.pgAdvisory});

  static const none = LockLayers(fiducia: false, pgAdvisory: false);
  static const fiduciaOnly = LockLayers(fiducia: true, pgAdvisory: false);
  static const pgOnly = LockLayers(fiducia: false, pgAdvisory: true);
  static const both = LockLayers(fiducia: true, pgAdvisory: true);

  @override
  bool operator ==(Object other) =>
      other is LockLayers &&
      other.fiducia == fiducia &&
      other.pgAdvisory == pgAdvisory;

  @override
  int get hashCode => Object.hash(fiducia, pgAdvisory);
}

/// How the Postgres advisory lock is scoped.
enum PgScope {
  /// `pg_advisory_xact_lock` inside a transaction the routine opens; the work
  /// runs inside it; released at commit/rollback.
  transaction('transaction'),

  /// `pg_advisory_lock` / `pg_advisory_unlock` on one dedicated connection;
  /// no transaction is opened.
  session('session');

  final String wire;
  const PgScope(this.wire);

  static PgScope parse(String value) =>
      values.firstWhere((s) => s.wire == value);
}

/// One action in a plan. [wire] is the contract's `LockStep` value.
enum LockStep {
  fiduciaAcquire('fiducia.acquire'),
  fiduciaTryAcquire('fiducia.try_acquire'),
  fiduciaRelease('fiducia.release'),
  pgBegin('pg.begin'),
  pgAdvisoryXactLock('pg.advisory_xact_lock'),
  pgTryAdvisoryXactLock('pg.try_advisory_xact_lock'),
  pgCommit('pg.commit'),
  pgRollback('pg.rollback'),
  pgAdvisoryLock('pg.advisory_lock'),
  pgTryAdvisoryLock('pg.try_advisory_lock'),
  pgAdvisoryUnlock('pg.advisory_unlock'),
  work('work');

  final String wire;
  const LockStep(this.wire);

  static LockStep parse(String value) =>
      values.firstWhere((s) => s.wire == value);
}

/// The ordered actions for one `(layers, scope, wait)` tuple.
final class LockPlan {
  final LockLayers layers;
  final PgScope pgScope;
  final bool wait;
  final List<LockStep> steps;

  const LockPlan(
      {required this.layers,
      required this.pgScope,
      required this.wait,
      required this.steps});
}

/// Compute the plan. Pure; identical across every language slice. [wait]
/// blocks each layer up to its budget; `!wait` uses the non-blocking form of
/// each acquisition and fails fast with `contention`.
LockPlan plan(LockLayers layers, PgScope pgScope, bool wait) {
  LockStep pick(LockStep blocking, LockStep nonBlocking) =>
      wait ? blocking : nonBlocking;
  final steps = <LockStep>[
    if (layers.fiducia)
      pick(LockStep.fiduciaAcquire, LockStep.fiduciaTryAcquire),
    if (!layers.pgAdvisory)
      LockStep.work
    else if (pgScope == PgScope.session) ...[
      pick(LockStep.pgAdvisoryLock, LockStep.pgTryAdvisoryLock),
      LockStep.work,
      LockStep.pgAdvisoryUnlock,
    ] else ...[
      LockStep.pgBegin,
      pick(LockStep.pgAdvisoryXactLock, LockStep.pgTryAdvisoryXactLock),
      LockStep.work,
      LockStep.pgCommit,
    ],
    if (layers.fiducia) LockStep.fiduciaRelease,
  ];
  return LockPlan(
      layers: layers,
      pgScope: pgScope,
      wait: wait,
      steps: List.unmodifiable(steps));
}

File: locks/.vendor/.zed/oresoftware/ores-locks-and-leases/src/ts/src/plan.ts
/**
 * The lock plan: which layers, in which order, as data. Every runtime derives
 * its behavior from the same pure function; the matrix lives in
 * `conformance/cases/lock-plan.json`.
 */

/** Which coordination layers a routine engages. Both false is a deliberate pass-through. */
export interface LockLayers {
  /** The fiducia-cloud lease: outermost, cross-host, TTL-bounded, fenced. */
  readonly fiducia: boolean;
  /** A Postgres advisory lock: inner, single-database, released by Postgres. */
  readonly pgAdvisory: boolean;
}

export const LAYERS_NONE: LockLayers = { fiducia: false, pgAdvisory: false };
export const LAYERS_FIDUCIA_ONLY: LockLayers = { fiducia: true, pgAdvisory: false };
export const LAYERS_PG_ONLY: LockLayers = { fiducia: false, pgAdvisory: true };
export const LAYERS_BOTH: LockLayers = { fiducia: true, pgAdvisory: true };

/**
 * How the Postgres advisory lock is scoped. `transaction`:
 * `pg_advisory_xact_lock` inside a transaction the routine opens; the work
 * runs inside it; released at commit/rollback. `session`:
 * `pg_advisory_lock` / `pg_advisory_unlock` on one dedicated connection; no
 * transaction is opened.
 */
export type PgScope = "transaction" | "session";

/** One action in a plan. The contract's `LockStep` enum. */
export type LockStep =
  | "fiducia.acquire"
  | "fiducia.try_acquire"
  | "fiducia.release"
  | "pg.begin"
  | "pg.advisory_xact_lock"
  | "pg.try_advisory_xact_lock"
  | "pg.commit"
  | "pg.rollback"
  | "pg.advisory_lock"
  | "pg.try_advisory_lock"
  | "pg.advisory_unlock"
  | "work";

export const ALL_STEPS: readonly LockStep[] = [
  "fiducia.acquire",
  "fiducia.try_acquire",
  "fiducia.release",
  "pg.begin",
  "pg.advisory_xact_lock",
  "pg.try_advisory_xact_lock",
  "pg.commit",
  "pg.rollback",
  "pg.advisory_lock",
  "pg.try_advisory_lock",
  "pg.advisory_unlock",
  "work",
];

export interface LockPlan {
  readonly layers: LockLayers;
  readonly pgScope: PgScope;
  readonly wait: boolean;
  readonly steps: readonly LockStep[];
}

/**
 * Compute the plan. Pure; identical across every language slice. `wait`
 * blocks each layer up to its budget; `!wait` uses the non-blocking form of
 * each acquisition and fails fast with `contention`.
 */
export function plan(layers: LockLayers, pgScope: PgScope, wait: boolean): LockPlan {
  const steps: LockStep[] = [];
  if (layers.fiducia) steps.push(wait ? "fiducia.acquire" : "fiducia.try_acquire");
  if (!layers.pgAdvisory) {
    steps.push("work");
  } else if (pgScope === "session") {
    steps.push(wait ? "pg.advisory_lock" : "pg.try_advisory_lock", "work", "pg.advisory_unlock");
  } else {
    steps.push("pg.begin", wait ? "pg.advisory_xact_lock" : "pg.try_advisory_xact_lock", "work", "pg.commit");
  }
  if (layers.fiducia) steps.push("fiducia.release");
  return { layers, pgScope, wait, steps };
}
